### PR TITLE
Bluetooth: Add a possibility to disable bondable mode

### PR DESCRIPTION
**Pull request by MariuszSkamra**
This adds a new function to API to enable/disable bonding flag in SMP Pairing Request/Response.

### CODE REVIEW

File: subsys/bluetooth/host/smp.c [hunk 1]
```diff
@@ -243,6 +243,7 @@ static struct bt_smp_br bt_smp_br_pool[CONFIG_BT_MAX_CONN];
 #endif /* CONFIG_BT_BREDR */
 
 static struct bt_smp bt_smp_pool[CONFIG_BT_MAX_CONN];
+static bool non_bondable;
```
Comment (jhedberg) [inline subsys/bluetooth/host/smp.c]: Why the negative form instead of 'bondable'?

**Review by jhedberg — CHANGES_REQUESTED** [review]
Is this really needed as a runtime API? I'd have thought that a Kconfig would be sufficient.

**Comment by carlescufi** [comment]
I agree with @jhedberg. If we could do with a Kconfig option then I'd honestly prefer that too.

**Comment by jhedberg** [comment]
@carlescufi for the automated tester that uses a single image a runtime API is by far the most convenient. What I'd do is add a Kconfig to set the default, and then provide this API to change it at runtime if the app really needs it (but clearly document that it should only be needed in very rare circumstances).

**Comment by jhedberg** [comment]
And if we're sure that the only valid use case for the runtime API is for testing, then we could even put it behind CONFIG_BT_TESTING

File: subsys/bluetooth/host/smp.c [hunk 2]
```diff
@@ -2275,6 +2290,14 @@ static u8_t get_auth(u8_t auth)
 		auth |= BT_SMP_AUTH_MITM;
 	}
 
+#if defined(CONFIG_BT_TESTING)
+	if (bondable) {
+		auth |= BT_SMP_AUTH_BONDING;
+	} else {
+		auth &= ~(BT_SMP_AUTH_BONDING);
```
Comment (jhedberg) [inline subsys/bluetooth/host/smp.c]: minor: the () seem unnecessary

**Review by jhedberg — APPROVED** [review]
Looks generally good (one small comment but not critical enough for "request changes")

File: subsys/bluetooth/host/smp.c (hunk 2 above)
Comment (MariuszSkamra) [reply subsys/bluetooth/host/smp.c]: done. I just copied the code few lines above
		auth &= ~(BT_SMP_AUTH_MITM);
:)

File: include/bluetooth/testing.h [hunk 3]
```diff
@@ -82,6 +84,15 @@ int bt_test_mesh_lpn_group_remove(u16_t *groups, size_t groups_count);
...
 int bt_test_mesh_rpl_clear(void);
 
+/** Enable/disable bonding.
+ *
+ *  Set/clear the Bonding flag in the Authentication Requirements of
+ *  SMP Pairing Request/Response data. The flag is enabled by default.
+ *
+ *  @param enable Value allowing/disallowing to be bondable.
+ */
+void bt_test_set_bondable(bool enable);
```
Comment (carlescufi) [inline include/bluetooth/testing.h]: Might as well add this to the public API?

File: include/bluetooth/conn.h [hunk 4]
```diff
@@ -386,6 +386,15 @@ struct bt_conn_cb {
  */
 void bt_conn_cb_register(struct bt_conn_cb *cb);
 
+/** Enable/disable bonding.
+ *
+ *  Set/clear the Bonding flag in the Authentication Requirements of
+ *  SMP Pairing Request/Response data. The flag is enabled by default.
```
Comment (jhedberg) [inline include/bluetooth/conn.h]: The last sentence is not accurate, is it? The default depends on the Kconfig setting. You should also make it clear that for the vast majority of apps calling this shouldn't be needed. What was the end-product use case that made you decide that this is needed in a wider context than just for testing?

File: subsys/bluetooth/host/smp.c [hunk 5]
```diff
@@ -243,6 +249,7 @@ static struct bt_smp_br bt_smp_br_pool[CONFIG_BT_MAX_CONN];
 #endif /* CONFIG_BT_BREDR */
 
 static struct bt_smp bt_smp_pool[CONFIG_BT_MAX_CONN];
+static bool bondable = !!BT_SMP_AUTH_BONDING_FLAGS;
```
Comment (jhedberg) [inline subsys/bluetooth/host/smp.c]: `static bool bondable = IS_ENABLED(CONFIG_BT_BONDABLE);` might be more readable?

**Review by carlescufi — APPROVED** [review]
@MariuszSkamra great, thank you.

**Comment by carlescufi** [comment]
@jhedberg are you fine with the API being public?

**Review by jhedberg — CHANGES_REQUESTED** [review]

**Comment by jhedberg** [comment]
> @jhedberg are you fine with the API being public?

If there's a good rationale/use-case for it, sure. Aren't the testing APIs also public? (just that they require enabling CONFIG_BT_TESTING first)

File: include/bluetooth/conn.h (hunk 4 above)
Comment (jhedberg) [reply include/bluetooth/conn.h]: @MariuszSkamra I discussed this with @carlescufi on IRC and the consensus was to keep this public. So please just address the two other issues I brought up.

**Comment by MariuszSkamra** [comment]
@jhedberg I think @carlescufi has right, putting this to public API, because this is not related to testing only. The end-product might want to use it as well.

Updated

**Review by jhedberg — APPROVED** [review]

**Pull request merged**